adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,267 @@
+            #       see Datadog::Quantization::HTTP.url
+
+            if (implementation = config[:url_quantization])
+              implementation.call(request_uri_or_path_info)
+            else
+              request_uri_or_path_info
+            end
+          end
+
+          def base_url(rack_request)
+            if rack_request.respond_to?(:base_url)
+              rack_request.base_url
+            else
+              # Compatibility for older Rack versions
+              rack_request.url.chomp(rack_request.fullpath)
+            end
+          end
+
+          # e.g., "/webshop/articles/4?s=1"
+          def full_path(rack_request)
+            rack_request.fullpath
+          end
+
+          # catch exceptions that may be raised in the middleware chain
+          # Note: if a middleware catches an Exception without re raising,
+          # the Exception cannot be recorded here.
+          def record_and_reraise_error(error, request_span:)
+            request_span.status = OpenTelemetry::Trace::Status.new(
+              OpenTelemetry::Trace::Status::INTERNAL_ERROR,
+              description: error.to_s
+            )
+
+            # TODO: implement span.set_error? (this is a specification-level issue):
+            # request_span.set_error(error) unless request_span.nil?
+            #
+            raise error
+          end
+
+          def set_attributes_after_request(span, status, headers, _response)
+            span.status = OpenTelemetry::Trace::Status.http_to_status(status)
+            span.set_attribute('http.status_code', status)
+
+            # NOTE: if data is available, it would be good to do this:
+            # set_attribute('http.route', ...
+            # e.g., "/users/:userID?
+            span.set_attribute('http.status_text', ::Rack::Utils::HTTP_STATUS_CODES[status])
+
+            allowed_response_headers(headers).each { |k, v| span.set_attribute(k, v) }
+          end
+
+          # @return Hash
+          def allowed_request_headers(env)
+            {}.tap do |result|
+              self.class.allowed_rack_request_headers.each do |hash|
+                result[self.class.build_attribute_name('http.request.headers.', hash[:header])] = env[hash[:rack_header]] if env.key?(hash[:rack_header])
+              end
+            end
+          end
+
+          # @return Hash
+          def allowed_response_headers(headers)
+            return EMPTY_HASH if headers.nil?
+
+            {}.tap do |result|
+              self.class.allowed_response_headers.each do |hash|
+                key = headers.keys.find { |k| hash[:search_for_keys].find { |k2| k == k2 } } ||
+                      headers.keys.find { |k| hash[:search_for_keys].find { |k2| k.upcase == k2 } }
+
+                next unless key
+
+                # cache string for next time:
+                hash[:search_for_keys] |= [key]
+
+                result[hash[:attribute_name]] = headers[key]
+              end
+            end
+          end
```

Comment:
This optimization seems similarly problematic. Can't we just store an array of allowed_response headers and then a map of header name to attribute name. E.g. ['foo', 'bar']. After a match on 'foo' and a case insensitive match on 'BAR' the map would look like {'foo' => 'http.response.headers.foo', 'BAR' => 'http.response.headers.bar'}.